tensorpack/utils/nvml.py

```diff
@@ -0,0 +1,290 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+# File: nvml.py
+
+from ctypes import (byref, c_uint, c_ulonglong,
+                    CDLL, create_string_buffer, NVML_ERROR_FUNCTION_NOT_FOUND,
+                    NVML_ERROR_LIBRARY_NOT_FOUND, POINTER, Structure)
+import threading
+
+
+__all__ = ['NvidiaContext']
+
+
+_nvmlReturn_t = c_uint
+
+
+NVML_DEVICE_UUID_BUFFER_SIZE = 80
+NVML_DEVICE_NAME_BUFFER_SIZE = 64
+
+
+nvmlLib = None
+_lib_lock = threading.Lock()
+nvml_init_count = 0
+
+
+NvmlErrorCodes = {"0": "NVML_SUCCESS",
+                  "1": "NVML_ERROR_UNINITIALIZED",
+                  "2": "NVML_ERROR_INVALID_ARGUMENT",
+                  "3": "NVML_ERROR_NOT_SUPPORTED",
+                  "4": "NVML_ERROR_NO_PERMISSION",
+                  "5": "NVML_ERROR_ALREADY_INITIALIZED",
+                  "6": "NVML_ERROR_NOT_FOUND",
+                  "7": "NVML_ERROR_INSUFFICIENT_SIZE",
+                  "8": "NVML_ERROR_INSUFFICIENT_POWER",
+                  "9": "NVML_ERROR_DRIVER_NOT_LOADED",
+                  "10": "NVML_ERROR_TIMEOUT",
+                  "11": "NVML_ERROR_IRQ_ISSUE",
+                  "12": "NVML_ERROR_LIBRARY_NOT_FOUND",
+                  "13": "NVML_ERROR_FUNCTION_NOT_FOUND",
+                  "14": "NVML_ERROR_CORRUPTED_INFOROM",
+                  "15": "NVML_ERROR_GPU_IS_LOST",
+                  "16": "NVML_ERROR_RESET_REQUIRED",
+                  "17": "NVML_ERROR_OPERATING_SYSTEM",
+                  "18": "NVML_ERROR_LIB_RM_VERSION_MISMATCH",
+                  "999": "NVML_ERROR_UNKNOWN"}
+
+
+def NvmlCodeToString(i):
+    return NvmlErrorCodes[str(i)]
+
+
+class NvmlException(Exception):
+    def __init__(self, error_code):
+        super(NvmlException, self).__init__(error_code)
+        self.error_code = error_code
+
+    def __str__(self):
+        return NvmlCodeToString(self.error_code)
+
+
+def CheckNvmlReturn(ret):
+    if (ret != 0):
+        raise NvmlException(ret)
+    return ret
+
+
+class NVML(object):
+
+    def load(self):
+        global nvmlLib
```

Comment:
Not necessary to make it global, right?

Reply:
It think it is correct, I want a singleton-pattern here

Reply:
But everytime load() is called the so file gets loaded again. 
What's singleton is `_NVML = NVML()`. This so can just be an attribute.

Reply:
And even if it's made an attribute, is it correct to load the so everytime load() is called? Should it be cached?